### DISCUSSION TITLE
[27 marzo 2023] Complessità

### DISCUSSION THREAD

**Discussion by matypist** [sapienzastudentsnetwork/automi-calcolabilita-complessita#9] [Complessità]
![image](https://github.com/sapienzastudentsnetwork/acc2324/assets/69389461/afc3701b-5a3f-45b2-9d78-3868ebfff7fd)

**Comment by Exyss** [comment]
Per il primo punto, propongo la seguente soluzione:

![image](https://github.com/sapienzastudentsnetwork/acc2324/assets/65043207/bc61e0e0-c1e0-4ed0-9b33-57173fbaf13f)
![image](https://github.com/sapienzastudentsnetwork/acc2324/assets/65043207/bf6174c6-d066-4e14-a229-455d8ac647d8)

------

Per il secondo punto, propongo la seguente soluzione:

![image](https://github.com/sapienzastudentsnetwork/acc2324/assets/65043207/ae4e7eb9-c161-42c3-9f16-572ebaf79041)
![image](https://github.com/sapienzastudentsnetwork/acc2324/assets/65043207/27d6aa0a-f6c1-4130-b439-45489c395ee1)

**Reply by lord-marcel** [reply]
@Exyss nel punto 1), io non mi trovo in una cosa. Confrontando i tuoi appunti con il libro, ho notato che nel libro l'algoritmo costruisce un nodo x per ciascun letterale nella formula, laddove nei tuoi appunti c'è scritto di costruire un nodo x ed un nodo x-negato, per ciascun letterale x. 
Poi segue "3k nodi derivanti dalle k triple + k nodi aggiunti alla fine", ma se costruisci due nodi per ciascun letterale, a meno che non compaia il negato di ogni letterale nella formula, no ti ritrovi con più di 3k nodi?

**Reply by Exyss** [reply]
Sì, deve essere un nodo per letterale, probabilmente ho copia-incollato il testo da un altro esercizio simile e mi sono dimenticato di cambiarlo.

EDIT: Sistemato lo screenshot

**Reply by dahKidou** [reply]
@Exyss Domanda: ma essendo che sappiamo che CLIQUE è già NP-Completo, non è possibile ridurre HALFCLIQUE direttamente a CLIQUE simulando CLIQUE con <G,k> in input, dove k = |V|/2 ?

**Reply by Exyss** [reply]
Attenzione che per dimostrare che HalfClique sia NP-Completo bisogna ridurre Clique ad HalfClique, non il contrario. Detto questo, è sicuramente possibile ridurre Clique ad HalfClique in modo diretto, però la riduzione non è così immediata in quanto  vanno gestiti casi come "e se la clique già c'era ma ora è troppo piccola/grande?" "e se ho introdotto una clique per sbaglio?"

Probabilmente si può fare prendendo un'istanza <G,k> di Clique e trasformandola in un'istanza <G'> in questo modo:
- Se $k \geq \frac{|V(G|}{2}$ allora G' è la copia di G dove sono stati aggiunti $2k-|V(G|$ nodi con grado 0
- Se $k < \frac{|V(G)|}{2}$ allora G' è la copia di G dove sono stati aggiunti $|V(G|-2k$ nodi connessi ad ogni altro nodo del grafo

Non ho verificato la correttezza, è solo un'idea di getto che mi è venuta cercando di bilanciare k e il nuovo numero di vertici
